### DISCUSSION TITLE
Use with SP-API

### DISCUSSION THREAD

**Discussion by pedrogabe** [FantasticFiasco/aws-signature-version-4#578] [Q&A] [answered]
Hi! 
Has anybody use this to make requests to the Amazon Selling Partner API? 
Everything I try gives me the same response: "_The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details_"
I admit I don't have much experience with the SP-API. Any suggestions are welcome

**Comment by FantasticFiasco** [comment]
I don't have any experience myself using this the SP-API. Let's leave this discussion open and see if anyone has any experience of it.

**Comment by pedrogabe** [comment, marked as answer]
I have reached a solution for this problem by myself.
According to [Amazon SP-API documentation](https://developer-docs.amazon.com/sp-api/docs/include-a-user-agent-header-in-all-requests) a "User-Agent" header is required, and its value should comply with the following format:
`AppId/AppVersionId (Language=LanguageNameAndOptionallyVersion)`
I tried to add this header as a default request header of my HttpClient with a code line similar to this:
`client.DefaultRequestHeaders.Add(Amazon.Util.HeaderKeys.UserAgentHeader, "App/0.0 (Language=Csharp)");`

Yet, this `Add` method converted my string into two different values splitting it by the space character. Hence, the `CanonicalRequest` class included a comma in the canonical string that Amazon was not expecting.
I temporarily solved it by modifying the `CanonicalRequest.HeaderValueSeparator` value, though I understand this is not correct since the separator for the values should remain a comma as it is specified in the [AWS documentation](https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html)

If anybody knows a way of including the header value including space characters without it being split, I'll appreciate the tip. In conclussion, I see no real bugs in the code and it has turned out to be very useful.

**Reply by FantasticFiasco** [reply]
Thanks for analyzing the issue and finding the solution!

Without knowing anything about the SP-API, is there a way to uniquely identify requests to that API, like the service name?

**Reply by pedrogabe** [reply]
The service name is "execute-api", which doesn't sound very unique. I understand the possible host Headers are "sellingpartnerapi-na.amazon.com", "sellingpartnerapi-eu.amazon.com" and "sellingpartnerapi-fe.amazon.com", according to the endpoints listed in [the documentation](https://developer-docs.amazon.com/sp-api/docs/sp-api-endpoints).
I'll let you know if I find anything more reliable.

**Reply by mungojam** [reply]
@pedrogabe the user agent bug you found actually affects API Gateway calls too. I've raised https://github.com/FantasticFiasco/aws-signature-version-4/issues/1155 and also created a PR to fix it: https://github.com/FantasticFiasco/aws-signature-version-4/pull/1156
